tests/src/test/java/alluxio/collections/IndexedSetIntegrationTest.java

```diff
@@ -0,0 +1,259 @@
+    Set<Long> ids = new HashSet<>();
+    Set<Integer> sizes = new HashSet<>();
+
+    // Verify the size.
+    int expectedCount = 0;
+    while (it.hasNext()) {
+      TestInfo info = it.next();
+      ids.add(info.getId());
+      sizes.add(info.getSize());
+      expectedCount++;
+    }
+    Assert.assertEquals(expectedCount, mIndexedSet.size());
+
+    // Verify the size according to the id index.
+    int count = 0;
+    for (Long id : ids) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mIdIndex, id);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+
+    // Verify the size according to the size index.
+    count = 0;
+    for (Integer size : sizes) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mSizeIndex, size);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+  }
+
+  @Test
+  public void basicConcurrentTest() throws Exception {
+    List<Future<?>> futures = new ArrayList<>();
+    List<ConcurrentTask> addTasks = new ArrayList<>();
+    List<ConcurrentTask> removeTasks = new ArrayList<>();
+
+    addTasks.add(new ConcurrentAdd());
+    addTasks.add(new ConcurrentAdd());
+    addTasks.add(new ConcurrentAdd());
+    addTasks.add(new ConcurrentAdd());
+    removeTasks.add(new ConcurrentRemove());
+    removeTasks.add(new ConcurrentRemove());
+    removeTasks.add(new ConcurrentRemoveByField());
+    removeTasks.add(new ConcurrentRemoveByField());
+    for (ConcurrentTask task : addTasks) {
+      futures.add(mThreadPool.submit(task));
+    }
+    for (ConcurrentTask task : removeTasks) {
+      futures.add(mThreadPool.submit(task));
+    }
+
+    CommonUtils.sleepMs(TEST_CASE_DURATION_MS);
+    mStopThreads.set(true);
+    for (Future<?> future : futures) {
+      future.get();
+    }
+
+    // Calculate how many elements have been added or removed.
+    long added = 0;
+    for (ConcurrentTask task : addTasks) {
+      added += task.getCount();
+    }
+    long removed = 0;
+    for (ConcurrentTask task : removeTasks) {
+      removed += task.getCount();
+    }
+
+    Assert.assertEquals(mIndexedSet.size(), added - removed);
+    verifySet();
+  }
+
+  @Test
+  public void concurrentTest() throws Exception {
+    List<Future<?>> futures = new ArrayList<>();
+
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
+    futures.add(mThreadPool.submit(new ConcurrentRemove()));
+    futures.add(mThreadPool.submit(new ConcurrentRemove()));
+    futures.add(mThreadPool.submit(new ConcurrentRemoveByField()));
+    futures.add(mThreadPool.submit(new ConcurrentRemoveByField()));
+    futures.add(mThreadPool.submit(new ConcurrentRemoveByIterator()));
```

Comment:
The nice thing about the basicConcurrentTest is that all the adds and removes return success or failure, so we can compute exactly how many elements should be in the set. I thought this was a useful test case.

With this test, we do all the operations, and the iterator delete does not tell us if the delete succeeded or not, so we can no longer test the exact counts.